pennylane_lightning/core/src/simulators/lightning_tensor/tncuda/MPSTNCuda.hpp

```diff
@@ -232,6 +236,73 @@ class MPSTNCuda final : public TNCudaBase<Precision, MPSTNCuda<Precision>> {
         }
     };
 
+    /**
+     * @brief Apply an MPO operator with the gate's MPO decomposition data
+     * provided by the user to the compute graph.
+     *
+     * This API only works for the MPS backend.
+     *
+     * @param tensors The MPO representation of a gate. Each element in the
+     * outer vector represents a MPO tensor site.
+     * @param wires The wire indices of the gate acts on. The size of this
+     * vector should match the size of the `tensors` vector.
+     * @param max_mpo_bond_dim The maximum bond dimension of the MPO operator.
+     */
+    void applyMPOOperation(const std::vector<std::vector<ComplexT>> &tensors,
+                           const std::vector<std::size_t> &wires,
+                           const std::size_t max_mpo_bond_dim) {
+        PL_ABORT_IF_NOT(
+            tensors.size() == wires.size(),
+            "The number of tensors should be equal to the number of "
+            "wires.");
+
+        // Create a queue of wire pairs to apply SWAP gates and MPO local target
+        // wires
+        const auto [local_wires, swap_wires_queue] =
+            create_swap_wire_pair_queue(wires);
+
+        // Apply SWAP gates to ensure the following MPO operator targeting at
+        // local wires
+        if (swap_wires_queue.size() > 0) {
+            for (const auto &swap_wires : swap_wires_queue) {
+                for (const auto &wire_pair : swap_wires) {
+                    BaseType::applyOperation("SWAP", wire_pair, false);
+                }
```

Comment:
Do you think you can use a std::for_each here?

Reply:
Nice one!